File: sdks/local-rp/java/src/test/java/community/catalyst/linkkeys/localrp/ClaimsConformanceTest.java
package community.catalyst.linkkeys.localrp;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;

import community.catalyst.linkkeys.localrp.testutil.Fixtures;
import community.catalyst.linkkeys.localrp.testutil.MiniJson.JsonValue;
import community.catalyst.linkkeys.localrp.wire.Cbor;
import community.catalyst.linkkeys.localrp.wire.Codec;
import community.catalyst.linkkeys.localrp.wire.Types.Claim;
import community.catalyst.linkkeys.localrp.wire.Types.ClaimSignature;
import community.catalyst.linkkeys.localrp.wire.Types.DomainPublicKey;
import community.catalyst.linkkeys.localrp.wire.Types.LocalRpTicketRedemptionResponse;

/**
 * Conformance vectors: {@code claims.json} &mdash; {@code Claim} wire
 * encoding and claim-signature verification (design doc's discriminator for
 * the {@code claim_value} bstr-vs-tstr trap: an SDK that wires it as text
 * passes its own self-tests and only these cross-implementation vectors
 * expose the bug). Every case is driven through this SDK's own verification
 * path ({@link Claims}), the same path {@link Complete} uses.
 */
class ClaimsConformanceTest {

    private static DomainPublicKey parseDomainKey(JsonValue k) {
        return new DomainPublicKey(
                k.get("key_id").asString(),
                Fixtures.hex(k.get("public_key_hex").asString()),
                k.get("fingerprint_hex").asString(),
                k.get("algorithm").asString(),
                k.get("key_usage").asString(),
                k.get("created_at").asString(),
                k.get("expires_at").asString(),
                k.getStringOrNull("revoked_at"),
                null,
                null);
    }

    /** Group a flat {@code domain_keys} JSON array into {@link Claims.DomainKeySet}s by their {@code domain} field. */
    private static List<Claims.DomainKeySet> parseDomainKeySets(JsonValue domainKeysArray) {
        Map<String, List<DomainPublicKey>> byDomain = new LinkedHashMap<>();
        for (JsonValue k : domainKeysArray.asArray()) {
            byDomain.computeIfAbsent(k.get("domain").asString(), d -> new ArrayList<>()).add(parseDomainKey(k));
        }
        List<Claims.DomainKeySet> out = new ArrayList<>();
        for (Map.Entry<String, List<DomainPublicKey>> e : byDomain.entrySet()) {
            out.add(new Claims.DomainKeySet(e.getKey(), e.getValue()));
        }
        return out;
    }

    private static ClaimError.Kind expectedKind(String s) {
        return switch (s) {
            case "signature_invalid" -> ClaimError.Kind.SIGNATURE_INVALID;
            case "key_not_found" -> ClaimError.Kind.KEY_NOT_FOUND;
            default -> throw new IllegalArgumentException("unmapped expected_error: " + s);
        };
    }

    @Test
    void positiveCasesRoundTripByteExactlyAndVerify() {
        JsonValue d = Fixtures.load("claims.json");
        assertEquals("linkkeys-claim-v2", d.get("tag").asString());
        String fileSubjectDomain = d.get("subject_domain").asString();
        List<Claims.DomainKeySet> defaultDomainKeys = parseDomainKeySets(d.get("domain_keys"));

        List<JsonValue> cases = d.get("cases").asArray();
        assertEquals(3, cases.size());

        for (JsonValue c : cases) {
            String name = c.get("name").asString();
            assertTrue(c.get("expected_valid").asBoolean(), name);
            String subjectDomain = c.get("subject_domain").asString();
            assertEquals(fileSubjectDomain, subjectDomain, name);

            byte[] claimBytes = Fixtures.hex(c.get("claim_cbor_hex").asString());
            Claim claim = Codec.decodeClaim(claimBytes);
            assertArrayEquals(claimBytes, Codec.encodeClaim(claim), name + ": byte-exact re-encode");

            JsonValue claimJson = c.get("claim");
            assertEquals(claimJson.get("claim_id").asString(), claim.claimId(), name);
            assertEquals(claimJson.get("user_id").asString(), claim.userId(), name);
            assertEquals(claimJson.get("claim_type").asString(), claim.claimType(), name);
            assertArrayEquals(
                    Fixtures.hex(claimJson.get("claim_value_hex").asString()),
                    claim.claimValue(),
                    name + ": claim_value must be wire bytes (bstr), not CBOR text");
            assertEquals(claimJson.get("attested_at").asString(), claim.attestedAt(), name);
            assertEquals(claimJson.get("created_at").asString(), claim.createdAt(), name);
            assertEquals(claimJson.getStringOrNull("expires_at"), claim.expiresAt(), name);
            assertEquals(claimJson.getStringOrNull("revoked_at"), claim.revokedAt(), name);

            List<JsonValue> sigJsons = claimJson.get("signatures").asArray();
            assertEquals(sigJsons.size(), claim.signatures().size(), name);
            for (int i = 0; i < sigJsons.size(); i++) {
                JsonValue sigJson = sigJsons.get(i);
                ClaimSignature sig = claim.signatures().get(i);
                String signingDomain = sigJson.get("domain").asString();
                assertEquals(signingDomain, sig.domain(), name);
                assertEquals(sigJson.get("signed_by_key_id").asString(), sig.signedByKeyId(), name);
                assertArrayEquals(Fixtures.hex(sigJson.get("signature_hex").asString()), sig.signature(), name);

                // Re-derive the exact 8-element signed-payload array this SDK would sign/verify
                // over and confirm it matches the vector byte-for-byte.
                byte[] expectedPayload = Fixtures.hex(sigJson.get("signed_payload_cbor_hex").asString());
                byte[] computedPayload = Claims.claimSignPayload(
                        claim.claimId(),
                        claim.claimType(),
                        claim.claimValue(),
                        claim.userId(),
                        subjectDomain,
                        signingDomain,
                        claim.expiresAt(),
                        claim.attestedAt());
                assertArrayEquals(expectedPayload, computedPayload, name + ": signed_payload_cbor_hex mismatch");
            }

            // Through the SDK's own claim verification path, as the Complete flow uses.
            Claims.verifyClaim(claim, subjectDomain, defaultDomainKeys);
        }
    }

    @Test
    void negativeCasesFailWithExpectedErrorKind() {
        JsonValue d = Fixtures.load("claims.json");
        List<Claims.DomainKeySet> defaultDomainKeys = parseDomainKeySets(d.get("domain_keys"));
        List<JsonValue> cases = d.get("negative_cases").asArray();
        assertEquals(4, cases.size());

        for (JsonValue c : cases) {
            String name = c.get("name").asString();
            byte[] claimBytes = Fixtures.hex(c.get("claim_cbor_hex").asString());
            Claim claim = Codec.decodeClaim(claimBytes);
            String subjectDomain = c.get("subject_domain").asString();
            JsonValue override = c.getOrNull("domain_keys");
            List<Claims.DomainKeySet> domainKeys = override == null ? defaultDomainKeys : parseDomainKeySets(override);
            ClaimError.Kind expected = expectedKind(c.get("expected_error").asString());

            ClaimError err = assertThrows(
                    ClaimError.class,
                    () -> Claims.verifyClaimSignatures(claim, subjectDomain, domainKeys),
                    name + " unexpectedly verified");
            assertEquals(expected, err.kind(), name);
        }
    }

    @Test
    void decodeNegativeCaseRejectsClaimValueEncodedAsCborText() {
        JsonValue d = Fixtures.load("claims.json");
        List<JsonValue> cases = d.get("decode_negative_cases").asArray();
        assertEquals(1, cases.size());

        for (JsonValue c : cases) {
            String name = c.get("name").asString();
            assertFalse(c.get("expected_decode_ok").asBoolean(), name);
            byte[] claimBytes = Fixtures.hex(c.get("claim_cbor_hex").asString());
            assertThrows(
                    Cbor.CborDecodeException.class,
                    () -> Codec.decodeClaim(claimBytes),
                    name + ": a tstr claim_value must be rejected by a strict bstr codec");
        }
    }

    @Test
    void ticketRedemptionResponseRoundTripsAndEmbeddedClaimsVerify() {
        JsonValue d = Fixtures.load("claims.json");
        List<Claims.DomainKeySet> defaultDomainKeys = parseDomainKeySets(d.get("domain_keys"));
        JsonValue t = d.get("ticket_redemption_response");

        byte[] responseBytes = Fixtures.hex(t.get("response_cbor_hex").asString());
        LocalRpTicketRedemptionResponse resp = Codec.decodeLocalRpTicketRedemptionResponse(responseBytes);

        assertEquals(t.get("user_id").asString(), resp.userId());
        assertEquals(t.get("user_domain").asString(), resp.userDomain());
        assertEquals(t.get("ticket_expires_at").asString(), resp.ticketExpiresAt());
        assertEquals(3, resp.claims().size());

        assertArrayEquals(
                responseBytes,
                Codec.encodeLocalRpTicketRedemptionResponse(resp),
                "LocalRpTicketRedemptionResponse byte-exact re-encode");

        // Decoding without verifying fails the point: verify every embedded claim
        // through the same path complete_local_login's caller would use.
        for (Claim claim : resp.claims()) {
            Claims.verifyClaim(claim, resp.userDomain(), defaultDomainKeys);
        }
    }
}

File: sdks/local-rp/kotlin/src/test/kotlin/community/catalyst/linkkeys/localrp/kt/ClaimsConformanceTest.kt
package community.catalyst.linkkeys.localrp.kt

import java.time.OffsetDateTime
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import community.catalyst.linkkeys.localrp.kt.protocol.Claims
import community.catalyst.linkkeys.localrp.kt.protocol.ClaimDomainKeySet
import community.catalyst.linkkeys.localrp.kt.protocol.Crypto
import community.catalyst.linkkeys.localrp.kt.protocol.TicketRedemptionResponses
import community.catalyst.linkkeys.localrp.kt.testutil.Fixtures
import community.catalyst.linkkeys.localrp.kt.testutil.MiniJson.JsonValue

/**
 * Conformance vectors: `claims.json` -- claim wire encoding and
 * claim-signature verification, exercised through this package's own
 * [Claims] and [TicketRedemptionResponses].
 *
 * The trap this file exists to catch (README, `claims.json`): `claim_value`
 * is CBOR *bytes* (bstr), never text (tstr), both on the wire and inside the
 * signed payload -- a codec wired as tstr passes its own self-tests
 * (sign-wrong/verify-wrong is self-consistent) and only these
 * cross-implementation vectors expose it.
 */
class ClaimsConformanceTest {

    private fun parseDomainKey(k: JsonValue): DomainPublicKey = DomainPublicKey(
        k.get("key_id").asString(),
        Fixtures.hex(k.get("public_key_hex").asString()),
        k.get("fingerprint_hex").asString(),
        k.get("algorithm").asString(),
        k.get("key_usage").asString(),
        OffsetDateTime.parse(k.get("created_at").asString()).toInstant(),
        OffsetDateTime.parse(k.get("expires_at").asString()).toInstant(),
        k.getStringOrNull("revoked_at")?.let { OffsetDateTime.parse(it).toInstant() },
        null,
        null,
    )

    /** Group a flat `domain_keys[]` JSON array (each entry carries its own `domain` field) into per-domain key sets, as [Claims.verifyClaim]/[Claims.verifyClaimSignatures] require. */
    private fun parseDomainKeySets(domainKeysJson: JsonValue): List<ClaimDomainKeySet> =
        domainKeysJson.asArray()
            .groupBy { it.get("domain").asString() }
            .map { (domain, entries) -> ClaimDomainKeySet(domain, entries.map { parseDomainKey(it) }) }

    private fun publicKeyOf(domainKeysJson: JsonValue, keyId: String): ByteArray =
        Fixtures.hex(domainKeysJson.asArray().first { it.get("key_id").asString() == keyId }.get("public_key_hex").asString())

    @Test
    fun positiveCasesRoundTripAndVerify() {
        val d = Fixtures.load("claims.json")
        val defaultDomainKeys = parseDomainKeySets(d.get("domain_keys"))
        val cases = d.get("cases").asArray()
        assertEquals(3, cases.size, "expected 3 positive cases")

        for (c in cases) {
            val name = c.get("name").asString()
            assertTrue(c.get("expected_valid").asBoolean(), "$name: expected_valid should be true for cases[]")

            val claimCbor = Fixtures.hex(c.get("claim_cbor_hex").asString())
            val subjectDomain = c.get("subject_domain").asString()
            assertEquals("conformance.example", subjectDomain)

            // Byte-exact wire round-trip: decode then re-encode must reproduce
            // claim_cbor_hex exactly -- this is what catches a bstr/tstr mixup
            // for claim_value.
            assertArrayEquals(claimCbor, Claims.reencodeCbor(claimCbor), "$name: claim_cbor_hex did not round-trip byte-exactly")

            val claim = Claims.decode(claimCbor)
            val claimJson = c.get("claim")
            assertEquals(claimJson.get("claim_id").asString(), claim.claimId, name)
            assertEquals(claimJson.get("user_id").asString(), claim.userId, name)
            assertEquals(claimJson.get("claim_type").asString(), claim.claimType, name)
            assertArrayEquals(Fixtures.hex(claimJson.get("claim_value_hex").asString()), claim.claimValue, "$name: claim_value must be wired as bstr, not tstr")
            assertEquals(OffsetDateTime.parse(claimJson.get("attested_at").asString()).toInstant(), claim.attestedAt, name)
            assertEquals(OffsetDateTime.parse(claimJson.get("created_at").asString()).toInstant(), claim.createdAt, name)
            val expiresAtJson = claimJson.getStringOrNull("expires_at")
            if (expiresAtJson == null) {
                assertNull(claim.expiresAt, "$name: expires_at should be absent")
            } else {
                assertEquals(OffsetDateTime.parse(expiresAtJson).toInstant(), claim.expiresAt, name)
            }
            assertNull(claim.revokedAt, "$name: revoked_at should be absent")

            val sigsJson = claimJson.get("signatures").asArray()
            assertEquals(sigsJson.size, claim.signatures.size, "$name: signature count mismatch")
            for (i in sigsJson.indices) {
                val sigJson = sigsJson[i]
                val sig = claim.signatures[i]
                assertEquals(sigJson.get("domain").asString(), sig.domain, name)
                assertEquals(sigJson.get("signed_by_key_id").asString(), sig.signedByKeyId, name)
                assertArrayEquals(Fixtures.hex(sigJson.get("signature_hex").asString()), sig.signature, name)

                // Independent, lower-level cross-check: Ed25519-verify the
                // vector's own signed_payload_cbor_hex/signature_hex directly
                // against the signer's public key, without going through
                // Claims.verifyClaim at all.
                val payload = Fixtures.hex(sigJson.get("signed_payload_cbor_hex").asString())
                val signature = Fixtures.hex(sigJson.get("signature_hex").asString())
                val publicKey = publicKeyOf(d.get("domain_keys"), sigJson.get("signed_by_key_id").asString())
                assertTrue(Crypto.verifyEd25519(payload, signature, publicKey), "$name: raw Ed25519 verify of signed_payload_cbor_hex failed")
            }

            // Full verification (signature quorum + revocation/expiry) through
            // this package's own public surface.
            assertDoesNotThrow({ Claims.verifyClaim(claimCbor, subjectDomain, defaultDomainKeys) }, "$name unexpectedly failed to verify")
            assertDoesNotThrow({ Claims.verifyClaimSignatures(claimCbor, subjectDomain, defaultDomainKeys) }, "$name unexpectedly failed signature-only verification")
        }
    }

    @Test
    fun decodeNegativeCasesRejected() {
        val d = Fixtures.load("claims.json")
        val cases = d.get("decode_negative_cases").asArray()
        assertEquals(1, cases.size)

        for (c in cases) {
            val name = c.get("name").asString()
            assertFalse(c.get("expected_decode_ok").asBoolean(), name)
            val claimCbor = Fixtures.hex(c.get("claim_cbor_hex").asString())
            // claim_value_as_cbor_text_rejected: byte-identical to a positive
            // case except claim_value is encoded as CBOR major type 3 (text)
            // instead of major type 2 (bytes). A strict bstr codec must
            // refuse to decode this at all.
            assertThrows(RuntimeException::class.java, { Claims.decode(claimCbor) }, "$name unexpectedly decoded")
        }
    }

    @Test
    fun negativeCasesFailVerification() {
        val d = Fixtures.load("claims.json")
        val defaultDomainKeys = parseDomainKeySets(d.get("domain_keys"))
        val cases = d.get("negative_cases").asArray()
        assertEquals(4, cases.size)

        for (c in cases) {
            val name = c.get("name").asString()
            val claimCbor = Fixtures.hex(c.get("claim_cbor_hex").asString())
            val subjectDomain = c.get("subject_domain").asString()
            val expectedErrorName = c.get("expected_error").asString()
            val expectedKind = ClaimErrorKind.valueOf(expectedErrorName.uppercase())

            val domainKeysOverride = c.getOrNull("domain_keys")
            val domainKeys = if (domainKeysOverride != null) parseDomainKeySets(domainKeysOverride) else defaultDomainKeys

            val ex = assertThrows(LocalRpException.ClaimVerification::class.java, { Claims.verifyClaim(claimCbor, subjectDomain, domainKeys) }, "$name unexpectedly verified")
            assertEquals(expectedKind, ex.kind, "$name: wrong ClaimErrorKind")
        }
    }

    @Test
    fun ticketRedemptionResponseRoundTripsAndClaimsVerify() {
        val d = Fixtures.load("claims.json")
        val defaultDomainKeys = parseDomainKeySets(d.get("domain_keys"))
        val respJson = d.get("ticket_redemption_response")
        val responseCbor = Fixtures.hex(respJson.get("response_cbor_hex").asString())

        // Byte-exact whole-message round-trip.
        assertArrayEquals(responseCbor, TicketRedemptionResponses.reencodeCbor(responseCbor), "response_cbor_hex did not round-trip byte-exactly")

        val response = TicketRedemptionResponses.decode(responseCbor)
        assertEquals(respJson.get("user_id").asString(), response.userId)
        assertEquals(respJson.get("user_domain").asString(), response.userDomain)
        assertEquals(OffsetDateTime.parse(respJson.get("ticket_expires_at").asString()).toInstant(), response.ticketExpiresAt)
        assertEquals(3, response.claims.size, "expected all three positive-case claims, in order")

        val expectedNamesInOrder = listOf("claim_utf8_text_value", "claim_non_utf8_binary_value", "claim_multiple_signatures")
        val positiveCasesByName = d.get("cases").asArray().associateBy { it.get("name").asString() }
        val claimIdsInOrder = expectedNamesInOrder.map { positiveCasesByName.getValue(it).get("claim").get("claim_id").asString() }
        assertEquals(claimIdsInOrder, response.claims.map { it.claimId })

        // Decoding without verifying fails the point: verify every embedded
        // claim's signatures, from the claims' own re-encoded CBOR bytes
        // (never through the lossy Instant-backed Claim type -- see Claims'
        // class docs).
        val claimCborBytesList = TicketRedemptionResponses.claimCborBytes(responseCbor)
        assertEquals(3, claimCborBytesList.size)
        for (claimCbor in claimCborBytesList) {
            assertDoesNotThrow({ Claims.verifyClaim(claimCbor, response.userDomain, defaultDomainKeys) }, "embedded claim unexpectedly failed to verify")
        }
    }
}
